src/xo-app/vm/tab-disks.js

```diff
@@ -1,146 +1,481 @@
+})
+class AttachDisk extends Component {
+  constructor (props) {
+    super(props)
+    this.state = {}
+  }
+
+  _vdiPredicate = vdi => {
+    const { vm } = this.props
+    return vdi.$pool === vm.$pool
+  }
+
+  _srPredicate = sr => {
+    const { vm } = this.props
+    return sr.$pool === vm.$pool && sr.SR_type !== 'iso'
+  }
+
+  _selectVdi = vdi => this.setState({vdi})
+
+  _addVdi = () => {
+    const { vm, vbds, vdis, onClose = noop } = this.props
+    const { vdi } = this.state
+    const { bootable, readOnly } = this.refs
+    const _isFreeForWriting = (vdi) => {
+      let free = true
+      forEach(vdi.$VBDs, id => {
+        const vbd = vbds[id]
+        free = free && (!vbd || !vbd.attached || vbd.read_only)
+      })
+      return free
+    }
+    let lastPos = 0
+    forEach(vbds, vbd => {
+      if (vdis[vbd.VDI]) {
+        lastPos = Math.max(lastPos, +vbd.position)
+      }
+    })
+    const mode = readOnly.value || !_isFreeForWriting(vdi) ? 'RO' : 'RW'
+    return attachDiskToVM(vm, vdi, {
+      bootable: bootable.value,
+      mode,
+      position: lastPos + 1
+    })
+      .then(onClose)
+  }
+
+  render () {
+    const { vdi } = this.state
+    return <form id='attachDiskForm'>
+      <div className='form-group'>
+        <SelectVdi
+          predicate={this._vdiPredicate}
+          containerPredicate={this._srPredicate}
+          onChange={this._selectVdi}
+        />
+      </div>
+      {vdi && <fieldset className='form-inline'>
+        <div className='form-group'>
+          Bootable <Toggle ref='bootable' /> Readonly <Toggle ref='readOnly' />
+        </div>
+        <span className='pull-right'>
+          <ActionButton icon='add' form='attachDiskForm' btnStyle='primary' handler={this._addVdi}>Create</ActionButton>
+        </span>
+      </fieldset>
       }
-    </Col>
-  </Row>
-</Container>
+    </form>
+  }
+}
+
+const orderItemSource = {
+  beginDrag: props => ({
+    id: props.id,
+    index: props.index
+  })
+}
+
+const orderItemTarget = {
+  hover: (props, monitor, component) => {
+    const dragIndex = monitor.getItem().index
+    const hoverIndex = props.index
+
+    if (dragIndex === hoverIndex) {
+      return
+    }
+
+    props.move(dragIndex, hoverIndex)
+    monitor.getItem().index = hoverIndex
+  }
+}
+
+@DropTarget('orderItem', orderItemTarget, connect => ({
+  connectDropTarget: connect.dropTarget()
+}))
+@DragSource('orderItem', orderItemSource, (connect, monitor) => ({
+  connectDragSource: connect.dragSource(),
+  isDragging: monitor.isDragging()
+}))
+@propTypes({
+  connectDragSource: propTypes.func.isRequired,
+  connectDropTarget: propTypes.func.isRequired,
+  index: propTypes.number.isRequired,
+  isDragging: propTypes.bool.isRequired,
+  id: propTypes.any.isRequired,
+  item: propTypes.object.isRequired,
+  move: propTypes.func.isRequired
+})
+class OrderItem extends Component {
+  _toggle = checked => {
+    const { item } = this.props
+    item.active = checked
```

Comment:
Well, I chose to give to OrderItem the ability to work "by reference" on the item. Powerful sometimes, risky with widely shared objects, but as I said, this one is not widely shared, it is created just for the OrderItem to work with it.